text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,207 @@
+
+Implementation options:
+- Only direct queries (and endorsement requests) to the peer(s) with the highest ledger height.
+- Client proposals include last observed ledger height so Gateway can target any peers with at least that ledger height.
+ A view of ledger height can be tracked by the Gateway from block events it has received.
+
+## Scaling and load balancing
+
+Currently a client application is responsible for load balancing its requests between multiple peers.  The Gateway will handle this on behalf of the set of currently connected client applications.
+
+Ideally, the Gateway itself will be stateless (i.e. not maintain client session state) and so can be scaled independently with clients routing though an appriopriate load balancer.
+
+## Gateway SDKs
+
+A set of SDKs will be created to allow client applications to interact with the Fabric Gateway.  The APIs exposed by these SDKs will be, where possible, the same as the current high-level 'gateway' SDKs.
+
+The following classes/structures will be available as part of the gateway programming model:
+- Gateway
+- Network
+- Contract
+- Transaction
+- Wallet
+
+The existing `submitTransaction()` and `evaluateTransaction()` API methods will be implemented using the Gateway GRPC services described above.
+
+Based on user feedback of the existing gateway SDKs, the following additions are planned to be supported by this implementation:
+- Support for 'offline signing', allowing clients to write REST servers that don't directly have access to users' private keys.
+- Support for 'asynchronous submit', allowing client applications to continue processing while the transaction commit proceeds in the background. For Go this will be implemented by returning a channel; for Node by returning a Promise; and for Java by returning a Future.
+
+SDKs will be created for the following languages:
+- Node (Typescript/Javascript)
+- Go
+- Java
+
+It is proposed that these SDKs will be maintained in the same GitHub respository as the gateway itself (`hyperledger/fabric-gateway`) to ensure that they all stay up to date with each other and with the core gateway.  This will be enforced in the CI pipeline by running the end-to-end scenario tests across all SDKs.
+
+Publication of SDK releases will be as follows:
+- Node SDK published to NPM.
+- Java SDK published to Maven Central.
+- Go SDK published to separate GitHub repo (e.g. fabric-gateway-go) and tagged.
+
+Contributors are invited to create SDKs for other languages, e.g. Python, Rust.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When run as a standalone process, the gateway would add another docker containter to an (arguably) already complex Fabric architecture.  This could be mitigated by embedding the gateway server within the organisation's existing peers.  This feature will be designed to be embeddable, if required.
```

Comment:
This message is conveyed at least 3/4 times in the RFC. To me, this looks like a major decision and cannot be left out for later to decide during the implementation. IMO, we need to be explicit on the RFC. As the goal is to keep the peer as small as possible, I would prefer keeping the gateway outside the peer. Adding another process does not impact the usability as the entry point is going to be the gateway for clients. I would prefer micro-service architecture which can help in scalability and maintainability over a monolithic peer process with too many features packed into a single entity. Further, I see gateway as an advanced version of nginx proxy or pgbouncer given it does more than just load balancing. Hence, it is better to keep this add-on/optional feature outside the peer. 

It would be good to discuss the pros/cons of two approaches in the RFC itself and then make a decision.

Reply:
>  I would prefer micro-service architecture which can help in scalability and maintainability over a monolithic peer process with too many features packed into a single entity

Fabric is not a micro-service friendly system, no matter how hard you want it to be:

- It uses TLS pinning in the orderer
- It is not proxy friendly
- It has its own load balancing and failover logic, and attempts to put cloud-y mechanisms usually end up getting in Fabric's way

I actually am in favor of putting this inside the peer, as you can re-use existing authentication mechanisms which are Fabric native and you don't need to re-implement them, and also the downside that requires completely new scripts and configuration files to launch a brand new node type.

Reply:
As I said in the last line, pros/cons are associated with both approaches. I would still keep this gateway outside peer as I haven't see a strong reason for embedding this into peer. At a high level, I still do see this gateway for peer as pgbouncer for PostgreSQL. While coding for the ledger snapshot feature, at least I did feel that our dependency management is very complex and there are too many components to manage. Hence, only if it is absolute necessary, we need to embedd this into peer. If it can stay outside while adding some complexity to BaaS provider and duplicate certain part of the code, it is completely fine as it wouldn't create problem for us in the long term. This is my opinion.

Reply:
I agree with Yacov, this will be far simpler architecturally to embed this in the peer.  It won't add too much overhead.  But as a component, it can be designed so that it can be run standalone, if preferred.  I've added this as a discussion point in the RFC